go/teams/box_audit.go

```diff
@@ -0,0 +1,1100 @@
+	// same time.
+	jailMutex  sync.Mutex
+	queueMutex sync.Mutex
+
+	// The box audit jail has an LRU for performance, we need a mutex so we
+	// don't use a partially initialized jailLRU.
+	jailLRUMutex sync.Mutex
+	jailLRU      *lru.Cache
+}
+
+var _ libkb.TeamBoxAuditor = &BoxAuditor{}
+
+func (a *BoxAuditor) resetJailLRU() {
+	a.jailLRUMutex.Lock()
+	defer a.jailLRUMutex.Unlock()
+
+	jailLRU, err := lru.New(JailLRUSize)
+	// lru.New only returns an error on a negative size, so it's safe to panic
+	// on an error.
+	if err != nil {
+		panic(err)
+	}
+
+	a.jailLRU = jailLRU
+}
+
+func (a *BoxAuditor) getJailLRU() *lru.Cache {
+	a.jailLRUMutex.Lock()
+	defer a.jailLRUMutex.Unlock()
+	return a.jailLRU
+}
+
+func (a *BoxAuditor) OnLogout(mctx libkb.MetaContext) {
+	a.resetJailLRU()
+}
+
+func NewBoxAuditor(g *libkb.GlobalContext) *BoxAuditor {
+	a := &BoxAuditor{Version: CurrentBoxAuditVersion}
+	a.resetJailLRU()
+	return a
+}
+
+func NewBoxAuditorAndInstall(g *libkb.GlobalContext) {
+	if g.GetEnv().GetDisableTeamBoxAuditor() {
+		g.Log.CWarningf(context.TODO(), "Box auditor disabled: using dummy auditor")
+		g.SetTeamBoxAuditor(DummyBoxAuditor{})
+	} else {
+		g.SetTeamBoxAuditor(NewBoxAuditor(g))
+	}
+}
+
+func (a *BoxAuditor) initMctx(mctx libkb.MetaContext) libkb.MetaContext {
+	mctx = mctx.WithLogTag(BoxAuditTag)
+	mctx = mctx.WithCtx(context.WithValue(mctx.Ctx(), SkipBoxAuditCheckContextKey, true))
+	return mctx
+}
+
+// BoxAuditTeam performs one attempt of a BoxAudit. If one is in progress for
+// the teamid, make a new attempt. If exceeded max tries or hit a malicious
+// error, return a fatal error.  Otherwise, make a new audit and fill it with
+// one attempt. If the attempt failed nonfatally, enqueue it in the retry
+// queue. If it failed fatally, add it to the jail. If it failed for reasons
+// that are purely client-side, like a disk write error, we retry it as well
+// but distinguish it from a failure the server could have possibly maliciously
+// caused.
+func (a *BoxAuditor) BoxAuditTeam(mctx libkb.MetaContext, teamID keybase1.TeamID) (err error) {
+	mctx = a.initMctx(mctx)
+	defer mctx.TraceTimed(fmt.Sprintf("BoxAuditTeam(%s)", teamID), func() error { return err })()
+
+	if !ShouldRunBoxAudit(mctx) {
+		mctx.Debug("Box auditor feature flagged off; not auditing...")
+		return nil
+	}
+
+	lock := a.locktab.AcquireOnName(mctx.Ctx(), mctx.G(), teamID.String())
+	defer lock.Release(mctx.Ctx())
+	return a.boxAuditTeamLocked(mctx, teamID)
+}
+
+func (a *BoxAuditor) boxAuditTeamLocked(mctx libkb.MetaContext, teamID keybase1.TeamID) (err error) {
+	defer mctx.TraceTimed(fmt.Sprintf("boxAuditTeamLocked(%s)", teamID), func() error { return err })()
+
+	log, err := a.maybeGetLog(mctx, teamID)
+	if err != nil {
+		return ClientBoxAuditError{err}
+	}
+	if log == nil {
+		log = NewBoxAuditLog(a.Version)
+	}
+
+	isRetry := log.InProgress
+	rotateBeforeAudit := isRetry && !mctx.G().TestOptions.NoAutorotateOnBoxAuditRetry
```

Comment:
Maybe I'm misunderstanding what's going on here, but I'd maybe we should save "last seen key generation" somewhere in the `BoxAuditLog` struct and only generate if the generation hasn't changed? Otherwise someone might have rotated between our retry attempts and we might be doing unnecessary rotations.

Reply:
The audit still might fail if some reset since that rotate so it would have to be something like "audit, and if failed rotate and audit again". That could help, but I'm imagining that audits should never fail and it's unlikely that people will be auditing the same teams at the same time given how long the intervals are.

Reply:
Right, we don't expect the retry path to be usually taken so it makes sense to keep it simple here. Thanks!

Reply:
That being said I think there is something for keeping more data for auditing purposes, like the diff of the box summaries. Max wanted to keep it simple for now though, can easily add it in later